core/src/main/java/org/apache/calcite/plan/volcano/RelSubset.java

```diff
@@ -120,6 +121,16 @@
    */
   boolean triggerRule = false;
 
+  /**
+   * the upper bound of the last OptimizeGroup call
+   */
+  RelOptCost upperBound;
+
+  /**
+   * RelNode ids that is invoked passThrough method before
+   */
+  Set<Integer> passThroughCache;
```

Comment:
We can use RelNode here, instead of rel id, for debug-ability.

Reply:
Good Point. Will resolve it.